File: android/app/src/main/java/kz/qpvpn/vpn/WgProfile.kt
package kz.qpvpn.vpn

import kz.qpvpn.model.AppsMode

/** Разобранный профиль WireGuard. */
data class WgProfile(
    val privateKey: String,
    val addresses: List<String>,
    val dns: List<String>,
    val mtu: Int,
    val publicKey: String,
    val presharedKey: String,
    val endpoint: String,
    val keepalive: Int,
    /** Параметры маскировки AmneziaWG, если они были в файле. */
    val amneziaParams: Map<String, String> = emptyMap(),
) {
    /** Профиль сделан под AmneziaWG — обычный WireGuard такой сервер не примет. */
    val isAmnezia: Boolean get() = amneziaParams.isNotEmpty()

    val protocolName: String get() = if (isAmnezia) "AmneziaWG" else "WireGuard"

    val hasIpv6Address: Boolean get() = addresses.any { it.contains(":") }

    val endpointHost: String
        get() = endpoint.substringBeforeLast(':', "")

    /**
     * Собирает текст конфигурации для библиотеки WireGuard.
     *
     * AllowedIPs здесь — главное: именно этот список решает, что пойдёт
     * в туннель. Списки программ передаются теми же полями, что понимает
     * официальное приложение WireGuard.
     */
    fun toConfigText(
        allowedIps: List<String>,
        includeDns: Boolean,
        appsMode: AppsMode,
        apps: List<String>,
    ): String = buildString {
        appendLine("[Interface]")
        appendLine("PrivateKey = $privateKey")
        appendLine("Address = ${addresses.joinToString(", ")}")
        appendLine("MTU = $mtu")
        if (includeDns && dns.isNotEmpty()) {
            appendLine("DNS = ${dns.joinToString(", ")}")
        }
        // Параметры маскировки AmneziaWG: без них сервер Amnezia не ответит.
        for ((key, name) in AMNEZIA_FIELDS) {
            amneziaParams[key]?.let { appendLine("$name = $it") }
        }
        if (apps.isNotEmpty()) {
            when (appsMode) {
                AppsMode.ONLY_SELECTED -> appendLine("IncludedApplications = ${apps.joinToString(", ")}")
                AppsMode.EXCEPT_SELECTED -> appendLine("ExcludedApplications = ${apps.joinToString(", ")}")
                AppsMode.OFF -> Unit
            }
        }
        appendLine()
        appendLine("[Peer]")
        appendLine("PublicKey = $publicKey")
        if (presharedKey.isNotEmpty()) {
            appendLine("PresharedKey = $presharedKey")
        }
        appendLine("Endpoint = $endpoint")
        appendLine("AllowedIPs = ${allowedIps.joinToString(", ")}")
        if (keepalive > 0) {
            appendLine("PersistentKeepalive = $keepalive")
        }
    }

    companion object {

        class ParseError(message: String) : Exception(message)

        /** Имена параметров маскировки так, как их ждёт библиотека AmneziaWG. */
        private val AMNEZIA_FIELDS = listOf(
            "jc" to "Jc", "jmin" to "Jmin", "jmax" to "Jmax",
            "s1" to "S1", "s2" to "S2",
            "h1" to "H1", "h2" to "H2", "h3" to "H3", "h4" to "H4",
        )

        private fun isKey(value: String): Boolean =
            value.length == 44 && value.endsWith("=")

        /** Разбирает обычный файл .conf, который выдаёт сервер. */
        fun parse(text: String): WgProfile {
            var section = ""
            val iface = mutableMapOf<String, String>()
            val peer = mutableMapOf<String, String>()

            for (rawLine in text.lines()) {
                val line = rawLine.substringBefore('#').trim()
                if (line.isEmpty()) continue
                if (line.startsWith("[") && line.endsWith("]")) {
                    section = line.drop(1).dropLast(1).lowercase()
                    continue
                }
                val separator = line.indexOf('=')
                if (separator < 0) continue
                val key = line.substring(0, separator).trim().lowercase()
                val value = line.substring(separator + 1).trim()
                when (section) {
                    "interface" -> iface[key] = value
                    "peer" -> if (!peer.containsKey(key)) peer[key] = value
                }
            }

            if (iface.isEmpty()) throw ParseError("В конфиге нет секции [Interface].")
            if (peer.isEmpty()) throw ParseError("В конфиге нет секции [Peer].")

            val privateKey = iface["privatekey"].orEmpty()
            if (privateKey.isEmpty()) throw ParseError("В конфиге не хватает поля PrivateKey.")
            if (!isKey(privateKey)) throw ParseError("Приватный ключ имеет неверный формат.")

            val publicKey = peer["publickey"].orEmpty()
            if (publicKey.isEmpty()) throw ParseError("В конфиге не хватает поля PublicKey.")
            if (!isKey(publicKey)) throw ParseError("Публичный ключ сервера имеет неверный формат.")

            val endpoint = peer["endpoint"].orEmpty()
            if (endpoint.isEmpty()) throw ParseError("В конфиге не хватает поля Endpoint.")
            if (!endpoint.contains(':')) throw ParseError("Endpoint должен быть в виде адрес:порт.")

            val addresses = splitList(iface["address"].orEmpty())
            if (addresses.isEmpty()) throw ParseError("В конфиге не хватает поля Address.")

            val presharedKey = peer["presharedkey"].orEmpty()
            if (presharedKey.isNotEmpty() && !isKey(presharedKey)) {
                throw ParseError("Preshared-ключ имеет неверный формат.")
            }

            // Amnezia добавляет в [Interface] параметры маскировки: размеры
            // мусорных пакетов и подменённые заголовки. Обычный WireGuard их
            // не понимает, поэтому запоминаем отдельно.
            val amnezia = AMNEZIA_FIELDS.map { it.first }.mapNotNull { key ->
                iface[key]?.let { key to it }
            }.toMap()

            return WgProfile(
                privateKey = privateKey,
                addresses = addresses,
                dns = splitList(iface["dns"].orEmpty()),
                mtu = iface["mtu"]?.toIntOrNull()?.coerceIn(1200, 1500) ?: 1420,
                publicKey = publicKey,
                presharedKey = presharedKey,
                endpoint = endpoint,
                keepalive = peer["persistentkeepalive"]?.toIntOrNull() ?: 25,
                amneziaParams = amnezia,
            )
        }

        fun splitList(value: String): List<String> =
            value.split(',', ' ').map { it.trim() }.filter { it.isNotEmpty() }
    }
}

File: ios/Shared/WgProfile.swift
import Foundation

/// Разобранный профиль WireGuard или AmneziaWG.
struct WgProfile {

    /// Имена параметров маскировки так, как их ждёт библиотека AmneziaWG.
    static let amneziaFields: [(key: String, name: String)] = [
        ("jc", "Jc"), ("jmin", "Jmin"), ("jmax", "Jmax"),
        ("s1", "S1"), ("s2", "S2"), ("s3", "S3"), ("s4", "S4"),
        ("h1", "H1"), ("h2", "H2"), ("h3", "H3"), ("h4", "H4"),
        ("i1", "I1"), ("i2", "I2"), ("i3", "I3"), ("i4", "I4"), ("i5", "I5"),
    ]

    var privateKey: String
    var addresses: [String]
    var dns: [String]
    var mtu: Int
    var publicKey: String
    var presharedKey: String
    var endpoint: String
    var keepalive: Int

    /// Параметры маскировки AmneziaWG, если они были в файле.
    var amneziaParams: [String: String]

    var isAmnezia: Bool { !amneziaParams.isEmpty }
    var protocolName: String { isAmnezia ? "AmneziaWG" : "WireGuard" }

    var endpointHost: String {
        guard let colon = endpoint.lastIndex(of: ":") else { return endpoint }
        return String(endpoint[endpoint.startIndex..<colon])
    }

    enum ParseError: LocalizedError {
        case missing(String)

        var errorDescription: String? {
            switch self {
            case .missing(let what): return what
            }
        }
    }

    /// Собирает текст настроек для туннеля.
    ///
    /// AllowedIPs здесь — главное: именно этот список решает, что пойдёт
    /// в туннель.
    func configText(allowedIps: [String], includeDns: Bool) -> String {
        var lines: [String] = []
        lines.append("[Interface]")
        lines.append("PrivateKey = \(privateKey)")
        lines.append("Address = \(addresses.joined(separator: ", "))")
        lines.append("MTU = \(mtu)")
        if includeDns && !dns.isEmpty {
            lines.append("DNS = \(dns.joined(separator: ", "))")
        }
        for (key, name) in Self.amneziaFields {
            if let value = amneziaParams[key] {
                lines.append("\(name) = \(value)")
            }
        }
        lines.append("")
        lines.append("[Peer]")
        lines.append("PublicKey = \(publicKey)")
        if !presharedKey.isEmpty {
            lines.append("PresharedKey = \(presharedKey)")
        }
        lines.append("Endpoint = \(endpoint)")
        lines.append("AllowedIPs = \(allowedIps.joined(separator: ", "))")
        if keepalive > 0 {
            lines.append("PersistentKeepalive = \(keepalive)")
        }
        return lines.joined(separator: "\n") + "\n"
    }

    static func parse(_ text: String) throws -> WgProfile {
        var section = ""
        var interface: [String: String] = [:]
        var peer: [String: String] = [:]

        for rawLine in text.replacingOccurrences(of: "\r\n", with: "\n").split(separator: "\n", omittingEmptySubsequences: false) {
            var line = String(rawLine)
            if let hash = line.firstIndex(of: "#") {
                line = String(line[line.startIndex..<hash])
            }
            line = line.trimmingCharacters(in: .whitespaces)
            if line.isEmpty { continue }

            if line.hasPrefix("["), line.hasSuffix("]") {
                section = line.dropFirst().dropLast().lowercased()
                continue
            }

            guard let equals = line.firstIndex(of: "=") else { continue }
            let key = line[line.startIndex..<equals].trimmingCharacters(in: .whitespaces).lowercased()
            let value = line[line.index(after: equals)...].trimmingCharacters(in: .whitespaces)

            if section == "interface" { interface[key] = value }
            else if section == "peer" { peer[key] = value }
        }

        guard !interface.isEmpty else { throw ParseError.missing("В настройках нет раздела [Interface].") }
        guard !peer.isEmpty else { throw ParseError.missing("В настройках нет раздела [Peer].") }

        guard let privateKey = interface["privatekey"], !privateKey.isEmpty else {
            throw ParseError.missing("В настройках нет приватного ключа.")
        }
        guard let publicKey = peer["publickey"], !publicKey.isEmpty else {
            throw ParseError.missing("В настройках нет публичного ключа сервера.")
        }
        guard let endpoint = peer["endpoint"], !endpoint.isEmpty else {
            throw ParseError.missing("В настройках нет адреса сервера (Endpoint).")
        }

        var amnezia: [String: String] = [:]
        for (key, _) in amneziaFields {
            if let value = interface[key], !value.isEmpty { amnezia[key] = value }
        }

        return WgProfile(
            privateKey: privateKey,
            addresses: splitList(interface["address"]),
            dns: splitList(interface["dns"]),
            mtu: Int(interface["mtu"] ?? "") ?? 1420,
            publicKey: publicKey,
            presharedKey: peer["presharedkey"] ?? "",
            endpoint: endpoint,
            keepalive: Int(peer["persistentkeepalive"] ?? "") ?? 25,
            amneziaParams: amnezia
        )
    }

    private static func splitList(_ value: String?) -> [String] {
        (value ?? "")
            .split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
    }
}
